be/src/storage/tablet_updates.cpp

```diff
@@ -2370,10 +2372,69 @@ void TabletUpdates::_update_total_stats(const std::vector<uint32_t>& rowsets) {
     _cur_total_dels = ndel;
 }
 
-Status TabletUpdates::_get_column_values(std::vector<uint32_t>& column_ids, bool with_default,
-                                         std::map<uint32_t, std::vector<uint32_t>>& rowids_by_rssid,
-                                         vector<std::unique_ptr<vectorized::Column>>* columns) {
-    // TODO(cbl): impl
+Status TabletUpdates::get_column_values(std::vector<uint32_t>& column_ids, bool with_default,
+                                        std::map<uint32_t, std::vector<uint32_t>>& rowids_by_rssid,
+                                        vector<std::unique_ptr<vectorized::Column>>* columns) {
+    std::map<uint32_t, RowsetSharedPtr> rssid_to_rowsets;
+    {
+        std::lock_guard<std::mutex> l(_rowsets_lock);
+        for (const auto& rowset : _rowsets) {
+            rssid_to_rowsets.insert(rowset);
+        }
+    }
+    if (with_default) {
+        for (auto& column : *columns) {
+            column->append_default();
+        }
+    }
+    for (const auto& [rssid, rowids] : rowids_by_rssid) {
+        auto iter = rssid_to_rowsets.upper_bound(rssid);
+        --iter;
+        const auto& rowset = iter->second.get();
+        if (!(rowset->rowset_meta()->get_rowset_seg_id() <= rssid &&
+              rssid < rowset->rowset_meta()->get_rowset_seg_id() + rowset->num_segments())) {
+            string msg = Substitute("illegal rssid: $0, should in [$1, $2)", rssid,
+                                    rowset->rowset_meta()->get_rowset_seg_id(),
+                                    rowset->rowset_meta()->get_rowset_seg_id() + rowset->num_segments());
+            LOG(ERROR) << msg;
+            _set_error();
+            return Status::InternalError(msg);
+        }
+        std::string seg_path =
+                BetaRowset::segment_file_path(rowset->rowset_path(), rowset->rowset_id(), rssid - iter->first);
+        auto segment_res =
+                Segment::open(ExecEnv::GetInstance()->tablet_meta_mem_tracker(), fs::fs_util::block_manager(), seg_path,
+                              rssid - iter->first, &rowset->schema());
+        if (!segment_res.ok()) {
+            LOG(WARNING) << "Fail to open " << seg_path << ": " << segment_res.status();
+            return segment_res.status();
+        }
+        if ((*segment_res)->num_rows() == 0) {
```

Comment:
Why use *segment_res, not segment_res directly？

Reply:
Why num_rows() == 0 but not is_end_of_file()?

Reply:
> segment_res

the return value is statusor, not value directly, *segment_res is the value, segment_res is the status value

Reply:
> 

check segment num_row firstly, and then for latter segment iterator, use is_end_of_file.